crates/pairwise_nostd_internal/src/parallel.rs

```diff
@@ -0,0 +1,724 @@
+//! Our parallelism abstractions use the concepts of teams & reductions
+//!
+//! In general, a parallel reduction is a way of characterizing a certain kind
+//! of calculation that can be broken into parts, where each part can be
+//! computed simultaneously, and each of the partial results can be combined
+//! together into a single result.
+//!
+//! The idea of a "team" is an abstraction that we define in the context of
+//! describing parallelism. When we decompose a parallel reduction into pieces
+//! or "units of work":
+//! - we can distributed the units of work among 1 or more teams
+//! - a team is composed of 1 or more members. The members of a team work
+//!   together in a tightly-coupled, synchronous manner to collaboratively
+//!   complete a single task at a time.
+//!
+//! This abstraction nicely maps to hardware:
+//! - on GPUs, you can imagine that each team-member corresponds to a separate
+//!   thread in a single Warp (aka a WaveFront)
+//! - on CPUs, it is optimal for a team to be represented by a single thread,
+//!   and for each member to correspond to a lane use by SIMD instructions
+//!
+//! Of course we also have flexibility to adjust the definition of this
+//! hardware mapping
+
+use crate::accumulator::{Accumulator, Datum};
+use crate::reduce_utils::reset_full_statepack;
+use crate::state::StatePackViewMut;
+use core::num::NonZeroU32;
+
+/// Used to describes the properties of a team member.
+///
+/// The main reason purpose for this trait's existence is so that methods of
+/// [`ReductionSpec`] can accept a type that implements this trait as an
+/// argument and provide a specialization for:
+/// - the case when the members of a team are represented by a full thread
+/// - the case when the members of a team are vector-lanes driven by a single
+///   thread.
+///
+/// # Note
+/// If we are clever enough, I think we can probably get rid of this trait.
+pub trait TeamMemberProp {
+    const IS_VECTOR_PROCESSOR: bool;
+
+    /// when SELF::IS_VECTOR_PROCESSOR is `true`, this will return 0
+    fn get_id(&self) -> u32;
+}
```

Comment:
Why is this a trait (as opposed to a struct)?  Is there some other functionality that you would want WarpMember to have?

Reply:
This is a trait since we can specialize between a SIMD members and a ThreadMember (what you are calling a WarpMember).

It seems I had deleted the code corresponding to the case with SIMD members (probably because we weren't using it)